File: sky/org.kevoree.library.sky.api/src/main/scala/org/kevoree/library/sky/api/KevoreeNodeManager.scala
package org.kevoree.library.sky.api

import org.kevoree.library.sky.api.nodeType.KevoreeNodeRunnerFactory
import org.kevoree.ContainerRoot

import org.slf4j.{LoggerFactory, Logger}

/**
 * User: Erwan Daubert - erwan.daubert@gmail.com
 * Date: 22/09/11
 * Time: 11:46
 *
 * @author Erwan Daubert
 * @version 1.0
 */
class KevoreeNodeManager(kevoreeNodeRunnerFactory: KevoreeNodeRunnerFactory) {

  private val logger: Logger = LoggerFactory.getLogger(this.getClass)

  var runners = new scala.collection.mutable.ArrayBuffer[KevoreeNodeRunner] with scala.collection.mutable.SynchronizedBuffer[KevoreeNodeRunner]

  def stop() {
    logger.debug("try to stop all nodes")
    runners.foreach {
      runner => runner.stopNode()
    }
    runners = new scala.collection.mutable.ArrayBuffer[KevoreeNodeRunner] with scala.collection.mutable.SynchronizedBuffer[KevoreeNodeRunner]
  }

  def addNode(iaasModel: ContainerRoot, targetChildName: String, targetChildModel: ContainerRoot): Boolean = {
    logger.debug("try to add {}", targetChildName)
    val newRunner = kevoreeNodeRunnerFactory.createKevoreeNodeRunner(targetChildName)
    runners.append(newRunner)

    newRunner.addNode(iaasModel, targetChildModel)
  }


  def startNode(iaasModel: ContainerRoot, targetChildName: String, targetChildModel: ContainerRoot) : Boolean = {
    logger.debug("try to start {}", targetChildName)
    runners.find(runner => runner.nodeName == targetChildName) match {
      case None => true // we do nothing because there is no node with this name
      case Some(runner) => {
        runner.startNode(iaasModel, targetChildModel)
      }
    }
  }

  def stopNode(iaasModel: ContainerRoot, targetChildName: String): Boolean = {
    logger.debug("try to stop {}", targetChildName)
    runners.find(runner => runner.nodeName == targetChildName) match {
      case None => true // we do nothing because there is no node with this name
      case Some(runner) => {
        runner.stopNode()
      }
    }
  }

  def removeNode(iaasModel: ContainerRoot, targetChildName: String): Boolean = {
    logger.debug("try to remove {}", targetChildName)
    runners.find(runner => runner.nodeName == targetChildName) match {
      case None => true // we do nothing because there is no node with this name
      case Some(runner) => {
        runners.remove(runners.indexOf(runner))
        runner.removeNode()
      }
    }
  }
}

File: sky/org.kevoree.library.sky.api/src/main/java/org/kevoree/library/sky/api/execution/KevoreeNodeManager.java
package org.kevoree.library.sky.api.execution;

import org.kevoree.ContainerRoot;
import org.kevoree.library.sky.api.KevoreeNodeRunnerFactory;
import org.kevoree.log.Log;

import java.util.ArrayList;
import java.util.List;

/**
 * User: Erwan Daubert - erwan.daubert@gmail.com
 * Date: 22/09/11
 * Time: 11:46
 *
 * @author Erwan Daubert
 * @version 1.0
 */
public class KevoreeNodeManager {


    List<KevoreeNodeRunner> runners = new ArrayList<KevoreeNodeRunner>();

    private KevoreeNodeRunnerFactory kevoreeNodeRunnerFactory;

    public KevoreeNodeManager(KevoreeNodeRunnerFactory kevoreeNodeRunnerFactory) {
        this.kevoreeNodeRunnerFactory = kevoreeNodeRunnerFactory;
    }

    public void stop() {
        Log.debug("try to stop all nodes");

        for (KevoreeNodeRunner runner : runners) {
            runner.stopNode();
        }
        runners.clear();
    }

    public boolean addNode(ContainerRoot iaasModel, String targetChildName, ContainerRoot targetChildModel) {
        Log.debug("try to add {}", targetChildName);

        KevoreeNodeRunner newRunner = kevoreeNodeRunnerFactory.createKevoreeNodeRunner(targetChildName);
        runners.add(newRunner);

        return newRunner.addNode(iaasModel, targetChildModel);
    }


    public boolean startNode(ContainerRoot iaasModel, String targetChildName, ContainerRoot targetChildModel) {
        Log.debug("try to start {}", targetChildName);

        KevoreeNodeRunner runner = null;
        for (KevoreeNodeRunner r : runners) {
            if (r.getNodeName().equals(targetChildName)) {
                runner = r;
            }
        }
        if (runner != null) {
            return runner.startNode(iaasModel, targetChildModel);
        } else {
            return true;
        }
    }

    public boolean stopNode(ContainerRoot iaasModel, String targetChildName) {
        Log.debug("try to stop {}", targetChildName);

        KevoreeNodeRunner runner = null;
        for (KevoreeNodeRunner r : runners) {
            if (r.getNodeName().equals(targetChildName)) {
                runner = r;
            }
        }
        if (runner != null) {
            return runner.stopNode();
        } else {
            return true;
        }
    }

    public boolean removeNode(ContainerRoot iaasModel, String targetChildName) {
        Log.debug("try to remove {}", targetChildName);

        KevoreeNodeRunner runner = null;
        for (KevoreeNodeRunner r : runners) {
            if (r.getNodeName().equals(targetChildName)) {
                runner = r;
            }
        }
        if (runner != null) {
            return runner.removeNode();
        } else {
            return true;
        }
    }
}